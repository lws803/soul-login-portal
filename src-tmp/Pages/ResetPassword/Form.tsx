import { useState } from 'react';
import {
  FormControl,
  FormLabel,
  Input,
  FormErrorMessage,
  HStack,
  Button,
  Progress,
} from '@chakra-ui/react';
import axios from 'axios';
import { useFormik } from 'formik';
import * as Yup from 'yup';
import zxcvbn from 'zxcvbn';

import { BASE_URL } from '../../constants';

export default function Form({ token, setErrors, setIsSuccess }: Props) {
  const [isResetting, setIsResetting] = useState(false);

  const formik = useFormik({
    initialValues: {
      password: '',
    },
    onSubmit: (values) => {
      setErrors([]);
      setIsResetting(true);

      axios
        .post(`${BASE_URL}/v1/users/password-reset?token=${token}`, values)
        .then(() => setIsSuccess(true))
        .catch(({ response: { data, status } }) => {
          if (status === 400 && data.error === 'VALIDATION_ERROR') {
            formik.setErrors({
              password: 'Password is too weak!',
            });
          } else {
            setErrors([data.message]);
          }
        })
        .finally(() => {
          setIsResetting(false);
        });
    },
    validationSchema: Yup.object({
      password: Yup.string().required(),
    }),
    validateOnChange: false,
  });

  return (
    <form onSubmit={(e) => formik.handleSubmit(e as any)}>
      <FormControl
        isInvalid={!!formik.errors.password && formik.touched.password}
      >
        <FormLabel htmlFor="password">Password</FormLabel>
        {formik.values.password.length > 0 && (
          <Progress
            value={((zxcvbn(formik.values.password).score + 1) / 5) * 100}
            colorScheme="soul.green"
            size="xs"
            mb={1}
          />
        )}
        <Input
          id="password"
          name="password"
          type="password"
          onChange={formik.handleChange}
          value={formik.values.password}
          disabled={!token}
          aria-label="Password input"
          variant="filled"
          placeholder="Enter your new password"
        />
        {formik.errors.password && (
          <FormErrorMessage>{formik.errors.password}</FormErrorMessage>
        )}
      </FormControl>
      <HStack mt={8}>
        <Button
          type="submit"
          isLoading={isResetting}
          disabled={!token}
          bg="soul.pink.light"
          _hover={{ bg: 'soul.pink.dark' }}
          _active={{ bg: 'soul.pink.dark' }}
          width="100%"
        >
          Reset password
        </Button>
      </HStack>
    </form>
  );
}

type Props = {
  setErrors: (error: string[]) => void;
  setIsSuccess: (isSuccess: boolean) => void;
  token?: string | null;
};
