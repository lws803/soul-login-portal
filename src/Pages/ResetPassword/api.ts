import axios from 'axios';
import axiosRetry from 'axios-retry';

import { BASE_URL } from 'src/constants';
import { ApiError } from 'src/shared/apiTypes';

axiosRetry(axios, { retryDelay: axiosRetry.exponentialDelay, retries: 3 });

export const resetPassword = async ({
  token,
  newPassword,
}: {
  token: string;
  newPassword: string;
}) => {
  try {
    await axios.post(`${BASE_URL}/v1/users/password-reset?token=${token}`, {
      password: newPassword,
    });
    return { error: null };
  } catch ({ response: { data } }) {
    return { error: data as ApiError };
  }
};
