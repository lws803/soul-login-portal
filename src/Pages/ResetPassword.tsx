import { useState } from 'react';

import Form from './ResetPassword/Form';

import Status from '../components/Status';
import Page from '../components/Page';
import useQueryParams from '../hooks/useQueryParams';
import Title from '../components/Title';

export default function ResetPassword() {
  const query = useQueryParams();
  const [errors, setErrors] = useState<string[]>([]);
  const [isSuccess, setIsSuccess] = useState(false);

  return (
    <Page errors={errors} title={<Title title="Reset new password" />}>
      {isSuccess ? (
        <Status
          status="Password reset!"
          message="Success! Please close this page and sign in again."
        />
      ) : (
        <Form
          setErrors={setErrors}
          setIsSuccess={setIsSuccess}
          token={query.get('token')}
        />
      )}
    </Page>
  );
}
