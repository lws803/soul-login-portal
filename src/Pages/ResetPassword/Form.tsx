import {
  FormControl,
  FormLabel,
  Input,
  FormErrorMessage,
  HStack,
  Button,
  Progress,
} from '@chakra-ui/react';
import { useFormik } from 'formik';
import * as Yup from 'yup';
import zxcvbn from 'zxcvbn';

import { resetPassword } from './api';

export default function Form({ token, setErrors, setIsSuccess }: Props) {
  const formik = useFormik({
    initialValues: {
      password: '',
    },
    onSubmit: async (values) => {
      setErrors([]);
      const { error } = await resetPassword({
        token: token!,
        newPassword: values.password,
      });

      if (!error) setIsSuccess(true);
      if (error) {
        if (error.error === 'VALIDATION_ERROR') {
          formik.setErrors({
            password: 'Password is too weak!',
          });
        } else {
          setErrors([error.message]);
        }
      }
    },
    validationSchema: Yup.object({
      password: Yup.string().required(),
    }),
    validateOnChange: false,
  });

  return (
    <form onSubmit={(e) => formik.handleSubmit(e as any)}>
      <FormControl
        isInvalid={!!formik.errors.password && formik.touched.password}
      >
        <FormLabel htmlFor="password">Password</FormLabel>
        {formik.values.password.length > 0 && (
          <Progress
            value={((zxcvbn(formik.values.password).score + 1) / 5) * 100}
            colorScheme="soul.green"
            size="xs"
            mb={1}
          />
        )}
        <Input
          id="password"
          name="password"
          type="password"
          onChange={formik.handleChange}
          value={formik.values.password}
          disabled={!token}
          aria-label="Password input"
          variant="filled"
          placeholder="Enter your new password"
        />
        {formik.errors.password && (
          <FormErrorMessage>{formik.errors.password}</FormErrorMessage>
        )}
      </FormControl>
      <HStack mt={8}>
        <Button
          type="submit"
          isLoading={formik.isSubmitting}
          disabled={!token}
          bg="soul.pink.light"
          _hover={{ bg: 'soul.pink.dark' }}
          _active={{ bg: 'soul.pink.dark' }}
          width="100%"
        >
          Reset password
        </Button>
      </HStack>
    </form>
  );
}

type Props = {
  setErrors: (error: string[]) => void;
  setIsSuccess: (isSuccess: boolean) => void;
  token?: string | null;
};
