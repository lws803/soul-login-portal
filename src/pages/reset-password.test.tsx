import React from 'react';
import { render } from '@testing-library/react';
import { MemoryRouter } from 'react-router-dom';

import ResetPassword from './reset-password';

describe(ResetPassword, () => {
  const path = '/?token=CODE';

  it('renders', async () => {
    const { container } = render(
      <MemoryRouter initialEntries={[path]}>
        <ResetPassword />
      </MemoryRouter>,
    );

    expect(container).toMatchSnapshot();
  });

  it('disables button when code is not provided', async () => {
    const { getByText } = render(
      <MemoryRouter>
        <ResetPassword />
      </MemoryRouter>,
    );
    expect(getByText('Reset password')).toBeDisabled();
  });
});
