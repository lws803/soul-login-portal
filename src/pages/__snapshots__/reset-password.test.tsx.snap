// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ResetPassword renders 1`] = `
<div>
  <div
    class="css-y1gt6f"
  >
    <div
      class="css-gmuwbf"
    >
      <div
        class="css-3qxem8"
      >
        <form>
          <div
            class="chakra-form-control css-0"
            role="group"
          >
            <label
              class="chakra-form__label css-uvho5g"
              for="password"
              id="field-1-label"
            >
              Password
            </label>
            <input
              aria-label="Password input"
              class="chakra-input css-0"
              id="password"
              name="password"
              placeholder="Enter your new password"
              type="password"
              value=""
            />
          </div>
          <div
            class="chakra-stack css-nzla2i"
          >
            <button
              class="chakra-button css-a0tdue"
              type="submit"
            >
              Reset password
            </button>
          </div>
        </form>
        <div
          class="chakra-stack css-untcqu"
        />
      </div>
    </div>
    <div
      class="css-gmuwbf"
    >
      <div
        class="chakra-stack css-esy13k"
      >
        <a
          class="chakra-link css-epvm6"
          href="https://github.com/soul-project/login-portal"
          target="_blank"
        >
          Github
        </a>
        <a
          class="chakra-link css-epvm6"
          href="https://github.com/soul-project/soul"
          target="_blank"
        >
          Soul Network's Github
        </a>
        <a
          class="chakra-link css-epvm6"
          href="mailto:lws803@gmail.com"
          target="_blank"
        >
          Contact me
        </a>
      </div>
    </div>
  </div>
</div>
`;
